app/js/MigrationPersistor.js

```diff
@@ -0,0 +1,221 @@
+    } catch (err) {
+      if (err instanceof NotFoundError) {
+        const fallbackBucket = _getFallbackBucket(bucketName)
+        if (shouldCopy) {
+          return _copyFileFromFallback(
+            fallbackBucket,
+            bucketName,
+            key,
+            key,
+            true
+          )
+        } else {
+          return fallback.promises.getFileStream(fallbackBucket, key, opts)
+        }
+      }
+      throw err
+    }
+  }
+
+  async function _copyFromFallbackStreamAndVerify(
+    stream,
+    sourceBucket,
+    destBucket,
+    sourceKey,
+    destKey
+  ) {
+    try {
+      let sourceMd5
+      try {
+        sourceMd5 = await fallback.promises.getFileMd5Hash(
+          sourceBucket,
+          sourceKey
+        )
+      } catch (err) {
+        logger.warn(err, 'error getting md5 hash from fallback persistor')
+      }
+
+      await primary.promises.sendStream(destBucket, destKey, stream, sourceMd5)
+    } catch (err) {
+      const error = new WriteError({
+        message: 'unable to copy file to destination persistor',
+        info: {
+          sourceBucket,
+          destBucket,
+          sourceKey,
+          destKey
+        }
+      }).withCause(err)
+      metrics.inc('fallback.copy.failure')
+
+      try {
+        await primary.promises.deleteFile(destBucket, destKey)
+      } catch (err) {
+        error.info.cleanupError = new WriteError({
+          message: 'unable to clean up destination copy artifact',
+          info: {
+            destBucket,
+            destKey
+          }
+        }).withCause(err)
+      }
+
+      logger.warn({ error }, 'failed to copy file from fallback')
+      throw error
+    }
+  }
+
+  async function _copyFileFromFallback(
+    sourceBucket,
+    destBucket,
+    sourceKey,
+    destKey,
+    returnStream = false
+  ) {
+    metrics.inc('fallback.copy')
+    const sourceStream = await fallback.promises.getFileStream(
+      sourceBucket,
+      sourceKey,
+      {}
+    )
+
+    if (!returnStream) {
+      return _copyFromFallbackStreamAndVerify(
+        sourceStream,
+        sourceBucket,
+        destBucket,
+        sourceKey,
+        destKey
+      )
+    }
+
+    const tee = new Minipass()
+    const clientStream = new Minipass()
+    const copyStream = new Minipass()
+
+    tee.pipe(clientStream)
+    tee.pipe(copyStream)
+
+    // copy the file in the background
+    _copyFromFallbackStreamAndVerify(
+      copyStream,
+      sourceBucket,
+      destBucket,
+      sourceKey,
+      destKey
+    ).catch(
+      // the error handler in this method will log a metric and a warning, so
+      // we don't need to do anything extra here, but catching it will prevent
+      // unhandled promise rejection warnings
+      () => {}
+    )
+
+    // start piping the source stream into the tee after everything is set up,
+    // otherwise one stream may consume bytes that don't arrive at the other
+    sourceStream.pipe(tee)
```

Comment:
Is there a reason we can't do this with normal streams?  I am not sure why Minipass is needed.  If the code is executed as
```
sourceStream.pipe(clientStream)
sourceStream.pipe(copyStream)
```
there shouldn't be a problem with data being lost because the sourceStream will not start flowing until the next tick (based on the presence of the `process.nextTick` in the `resume` method [_stream_readable.js#L949](https://github.com/nodejs/node/blob/43fb6ffef7a14b47d1b0ba8134102795f5017a59/lib/_stream_readable.js#L949) which is called by the `pipe` method at [_stream_readable.js#L780](https://github.com/nodejs/node/blob/43fb6ffef7a14b47d1b0ba8134102795f5017a59/lib/_stream_readable.js#L780) ->  [_stream_readable.js#L940](https://github.com/nodejs/node/blob/43fb6ffef7a14b47d1b0ba8134102795f5017a59/lib/_stream_readable.js#L940)  )

Reply:
Right, so I read something somewhere when first doing some googling that nodejs streams don't handle multiple pipe destinations well if one is a lot slower than the other and can result in a buffer getting overwritten.

I cannot find this information again, so either it was very old or I dreamt it, but I have been labouring under the misapprehension that you should use a passthrough stream if you want to 'tee' stuff from a to b.

In fact, the top answer for nodejs tee streams explains that this is indeed a comon pattern.
![omg](https://user-images.githubusercontent.com/6530014/73640746-becdfa80-4666-11ea-9ddb-e3cb4a860b8d.gif)